src/main/java/org/terracotta/build/services/Git.java

```diff
@@ -0,0 +1,87 @@
+package org.terracotta.build.services;
+
+import org.gradle.api.Action;
+import org.gradle.api.GradleException;
+import org.gradle.api.Project;
+import org.gradle.api.provider.Property;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.process.ExecOperations;
+import org.gradle.process.ExecSpec;
+import org.gradle.process.internal.ExecException;
+import org.terracotta.build.ExecUtils;
+
+import javax.inject.Inject;
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.Arrays.asList;
+import static org.gradle.internal.Actions.composite;
+
+public abstract class Git implements BuildService<Git.Parameters> {
+  public interface Parameters extends BuildServiceParameters {
+    Property<String> getGitExecutable();
+  }
+
+  @Inject
+  public abstract ExecOperations getExecOperations();
+
+  public boolean hasLocalChange() {
+    return !execute(spec -> spec.args("status", "--porcelain=2")).isEmpty();
+  }
+
+  public String getCommitHash() {
+    return Optional.ofNullable(System.getenv("GIT_COMMIT"))
+        .orElseGet(() -> execute(spec -> spec.args("rev-parse", "HEAD")));
+  }
+
+  public String getBranch() {
+    return Optional.ofNullable(System.getenv("GIT_BRANCH"))
+        .orElseGet(() -> execute(spec -> spec.args("rev-parse", "--abbrev-ref", "HEAD")));
+  }
+
+  public String diff(String commitHash) {
+    return execute(spec -> spec.args("diff", commitHash, "."));
+  }
+
+  public String hash(String data) {
+    return execute(spec -> {
+      spec.args("hash-object", "--stdin");
+      spec.setStandardInput(new ByteArrayInputStream(data.getBytes(StandardCharsets.UTF_8)));
+    });
+  }
+
+  public String execute(Action<ExecSpec> action) {
+    return ExecUtils.execute(getExecOperations(), composite(spec -> {
+      spec.executable(getParameters().getGitExecutable().get());
+      spec.args("--no-pager");
```

Comment:
It feels like `--no-pager` shouldn't be needed because this should be a non-interactive shell?

Reply:
No this is required, at least with the diff command, otherwise it gives en empty output stream.